crates/marketplace/src/testing/integration.rs

```diff
@@ -0,0 +1,315 @@
+//! This module implements interfaces necessary to run marketplace builder
+//! in HotShot testing harness.
+
+use std::{collections::HashMap, fmt::Display, marker::PhantomData, sync::Arc, time::Duration};
+
+use async_compatibility_layer::art::async_spawn;
+use async_lock::RwLock;
+use async_trait::async_trait;
+use hotshot::types::SignatureKey;
+use hotshot_example_types::node_types::TestVersions;
+use hotshot_testing::{
+    block_builder::{BuilderTask, TestBuilderImplementation},
+    test_builder::BuilderChange,
+};
+use hotshot_types::{
+    data::{Leaf, ViewNumber},
+    message::UpgradeLock,
+    traits::{
+        block_contents::{vid_commitment, BlockPayload, EncodeBytes},
+        node_implementation::{ConsensusTime, NodeType},
+        signature_key::BuilderSignatureKey,
+        states::ValidatedState,
+    },
+};
+use marketplace_builder_shared::block::ParentBlockReferences;
+use tagged_base64::TaggedBase64;
+use url::Url;
+use vbs::version::StaticVersion;
+
+use crate::{
+    builder_state::BuilderState,
+    service::{
+        run_builder_service, BroadcastSenders, BuilderHooks, GlobalState, NoHooks, ProxyGlobalState,
+    },
+};
+
+const BUILDER_CHANNEL_CAPACITY: usize = 1024;
+
+/// Testing configuration for marketplace builder
+/// Stores hooks that will be used in the builder in a type-erased manner,
+/// allowing for runtime configuration of hooks to use in tests.
+struct TestMarketplaceBuilderConfig<Types>
+where
+    Types: NodeType,
+{
+    hooks: Arc<Box<dyn BuilderHooks<Types>>>,
+}
+
+impl<Types> Default for TestMarketplaceBuilderConfig<Types>
+where
+    Types: NodeType,
+{
+    fn default() -> Self {
+        Self {
+            hooks: Arc::new(Box::new(NoHooks(PhantomData))),
+        }
+    }
+}
+
+/// [`TestBuilderImplementation`] for marketplace builder.
+/// Passed as a generic parameter to [`TestRunner::run_test`], it is be used
+/// to instantiate builder API and builder task.
+struct MarketplaceBuilderImpl {}
+
+#[async_trait]
+impl<Types> TestBuilderImplementation<Types> for MarketplaceBuilderImpl
+where
+    Types: NodeType<View = ViewNumber>,
+    Types::InstanceState: Default,
+    for<'a> <<Types::SignatureKey as SignatureKey>::PureAssembledSignatureType as TryFrom<
+        &'a TaggedBase64,
+    >>::Error: Display,
+    for<'a> <Types::SignatureKey as TryFrom<&'a TaggedBase64>>::Error: Display,
+{
+    type Config = TestMarketplaceBuilderConfig<Types>;
+
+    /// This is mostly boilerplate to instantiate and start [`ProxyGlobalState`] APIs and initial [`BuilderState`]'s event loop.
+    /// [`BuilderTask`] it returns will be injected into consensus runtime by HotShot testing harness and
+    /// will forward transactions from hotshot event stream to the builder.
+    async fn start(
+        n_nodes: usize,
+        url: Url,
+        config: Self::Config,
+        _changes: HashMap<u64, BuilderChange>,
```

Comment:
Is this argument for future tests?

Reply:
It is required by the trait, in HotShot it is used to test how network behaves when some of the builders go offline / restart / start failing responses. As it mostly used to test proposer behavior rather than the builder's, I decided to ignore it here, although maybe we'll want to make use of it in some more complex tests